sklearn/utils/_tags.py

```diff
@@ -0,0 +1,67 @@
+import numpy as np
+
+_DEFAULT_TAGS = {
+    'non_deterministic': False,
+    'requires_positive_X': False,
+    'requires_positive_y': False,
+    'X_types': ['2darray'],
+    'poor_score': False,
+    'no_validation': False,
+    'multioutput': False,
+    "allow_nan": False,
+    'stateless': False,
+    'multilabel': False,
+    '_skip_test': False,
+    '_xfail_checks': False,
+    'multioutput_only': False,
+    'binary_only': False,
+    'requires_fit': True,
+    'preserves_dtype': [np.float64],
+    'requires_y': False,
+    'pairwise': False,
+}
+
+
+def _safe_tags(estimator, key=None):
+    """Safely get estimator tags.
+
+    :class:`~sklearn.BaseEstimator` provides the estimator tags machinery.
+    However, if an estimator does not inherit from this base class, we should
+    fall-back to the default tags.
+
+    For scikit-learn built-in estimators, we should still rely on
+    `self._get_tags()`. `_safe_tags(est)` should be used when we are not sure
+    where `est` comes from: typically `_safe_tags(self.base_estimator)` where
+    `self` is a meta-estimator, or in the common checks.
+
+    Parameters
+    ----------
+    estimator : estimator object
+        The estimator from which to get the tag.
+
+    key : str, default=None
+        Tag name to get. By default (`None`), all tags are returned.
+
+    Returns
+    -------
+    tags : dict or tag value
+        The estimator tags. A single value is returned if `key` is not None.
+    """
+    if hasattr(estimator, "_get_tags"):
+        tags_provider = "_get_tags()"
+        tags = estimator._get_tags()
+    elif hasattr(estimator, "_more_tags"):
+        tags_provider = "_more_tags()"
+        tags = {**_DEFAULT_TAGS, **estimator._more_tags()}
```

Comment:
Now that we rely on `_more_tags` regardless of inheritance, what's the rationale for defaulting to `_DEFAULT_TAGS` with `_more_tags` but not with `_get_tags`?

I admit I'm a bit lost on all the possible code paths and use-cases here. It seems that we're overly permissive in some cases while being restrictive in others, with no obvious reason. Things were clearer to me when the logic was "with inheritance -> define _more_tags, no inheritance -> define _get_tags".

 But anyway, feel free to merge if we need to move with the release. This is still experimental after all.

Reply:
I think the message is simpler to always recommend to define `_more_tags` for whether or not you inherit from `BaseEstimator`.

Reply:
The idea now always implements `_more_tags` and it will work and it should cover 99% of the use case.

The remaining 1% is no inheritance and people that want to use tags -> implement `_get_tags` with strong requirements on our side regarding defaults.